File: shared/types/document.ts
// Structured Document Model — shared contract for frontend and backend.
//
// Mirrors services/document_model.py. Every book is represented internally
// as a tree of typed, uniquely-identified nodes rather than plain text.
//
// Future modules (Editing, Images, Translation, DOCX, KDP Validator) all
// operate on this tree by node id.

export type DocumentNodeType =
  | "project"
  | "book"
  | "part"
  | "chapter"
  | "section"
  | "paragraph"
  | "sentence";

export interface DocumentNode {
  id: string;
  node_type: DocumentNodeType;
  title?: string | null;
  text?: string | null;
  position: number;
  status: string;
  kind?: string | null;
  metadata: Record<string, unknown>;
  attachments: Record<string, unknown>;
  parent_id?: string | null;
  children: DocumentNode[];
}

export interface StructuredDocument {
  project_id: string;
  book_id: string;
  root: DocumentNode;
}

// ---------------------------------------------------------------------------
// Per-level CRUD shapes
// ---------------------------------------------------------------------------

export interface PartCreateRequest {
  title: string;
  slug: string;
  position?: number;
  summary?: string | null;
  status?: string;
}

export interface PartUpdateRequest {
  title?: string;
  slug?: string;
  position?: number;
  summary?: string | null;
  status?: string;
}

export interface PartResponse {
  id: string;
  project_id: string;
  book_id: string;
  title: string;
  slug: string;
  position: number;
  summary: string | null;
  status: string;
  word_count: number;
  created_at: string;
  updated_at: string;
  chapter_count: number;
}

export interface ChapterCreateRequest {
  title: string;
  slug: string;
  part_id?: string | null;
  position?: number;
  summary?: string | null;
  status?: string;
}

export interface ChapterUpdateRequest {
  title?: string;
  slug?: string;
  part_id?: string | null;
  position?: number;
  summary?: string | null;
  status?: string;
}

export interface ChapterResponse {
  id: string;
  project_id: string;
  book_id: string;
  part_id: string | null;
  title: string;
  slug: string;
  position: number;
  summary: string | null;
  status: string;
  word_count: number;
  created_at: string;
  updated_at: string;
  section_count: number;
}

export interface SectionCreateRequest {
  title?: string | null;
  position?: number;
  status?: string;
}

export interface SectionUpdateRequest {
  title?: string | null;
  position?: number;
  status?: string;
}

export interface SectionResponse {
  id: string;
  project_id: string;
  book_id: string;
  chapter_id: string;
  title: string | null;
  position: number;
  status: string;
  word_count: number;
  created_at: string;
  updated_at: string;
  paragraph_count: number;
}

export interface ParagraphCreateRequest {
  kind?: string;
  position?: number;
  status?: string;
}

export interface ParagraphUpdateRequest {
  kind?: string;
  position?: number;
  status?: string;
}

export interface ParagraphResponse {
  id: string;
  project_id: string;
  book_id: string;
  chapter_id: string;
  section_id: string;
  kind: string;
  position: number;
  status: string;
  word_count: number;
  created_at: string;
  updated_at: string;
  sentence_count: number;
}

export interface SentenceCreateRequest {
  text: string;
  kind?: string;
  position?: number;
  status?: string;
}

export interface SentenceUpdateRequest {
  text?: string;
  kind?: string;
  position?: number;
  status?: string;
}

export interface SentenceResponse {
  id: string;
  project_id: string;
  book_id: string;
  chapter_id: string;
  section_id: string;
  paragraph_id: string;
  text: string;
  kind: string;
  position: number;
  status: string;
  created_at: string;
  updated_at: string;
}

File: backend/schemas/document.py
"""Pydantic schemas for the structured document hierarchy.

These mirror :class:`DocumentNode` for API boundaries and introduce level-
specific request/response models for CRUD operations on Parts, Chapters,
Sections, Paragraphs, and Sentences.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# DocumentNode — the canonical tree representation at the API boundary
# ---------------------------------------------------------------------------


class DocumentNodeSchema(BaseModel):
    """Serialized :class:`DocumentNode` for API responses.

    Future modules (Editing, Images, Translation, DOCX, KDP Validator) all
    accept this shape. It is a plain dict-tree, not an ORM model, so that
    frontend, services, and workers share a single contract.
    """

    id: UUID
    node_type: str = Field(..., pattern=r"^(project|book|part|chapter|section|paragraph|sentence)$")
    title: str | None = None
    text: str | None = None
    position: int = 0
    status: str = "draft"
    kind: str | None = None
    metadata: dict[str, object] = Field(default_factory=dict)
    attachments: dict[str, object] = Field(default_factory=dict)
    parent_id: UUID | None = None
    children: list[DocumentNodeSchema] = Field(default_factory=list)


class StructuredDocumentSchema(BaseModel):
    """API representation of a full book tree."""

    project_id: UUID
    book_id: UUID
    root: DocumentNodeSchema


# ---------------------------------------------------------------------------
# Per-level CRUD requests
# ---------------------------------------------------------------------------


class PartCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    slug: str = Field(min_length=1, max_length=320)
    position: int | None = None
    summary: str | None = None
    status: str = "draft"


class PartUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=300)
    slug: str | None = Field(default=None, min_length=1, max_length=320)
    position: int | None = None
    summary: str | None = None
    status: str | None = None


class PartResponse(BaseModel):
    id: UUID
    project_id: UUID
    book_id: UUID
    title: str
    slug: str
    position: int
    summary: str | None
    status: str
    word_count: int
    created_at: datetime
    updated_at: datetime
    chapter_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class ChapterCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    slug: str = Field(min_length=1, max_length=320)
    part_id: UUID | None = None
    position: int | None = None
    summary: str | None = None
    status: str = "draft"


class ChapterUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=300)
    slug: str | None = Field(default=None, min_length=1, max_length=320)
    part_id: UUID | None = None
    position: int | None = None
    summary: str | None = None
    status: str | None = None


class ChapterResponse(BaseModel):
    id: UUID
    project_id: UUID
    book_id: UUID
    part_id: UUID | None
    title: str
    slug: str
    position: int
    summary: str | None
    status: str
    word_count: int
    created_at: datetime
    updated_at: datetime
    section_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class SectionCreateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=300)
    position: int | None = None
    status: str = "draft"


class SectionUpdateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=300)
    position: int | None = None
    status: str | None = None


class SectionResponse(BaseModel):
    id: UUID
    project_id: UUID
    book_id: UUID
    chapter_id: UUID
    title: str | None
    position: int
    status: str
    word_count: int
    created_at: datetime
    updated_at: datetime
    paragraph_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class ParagraphCreateRequest(BaseModel):
    kind: str = "body"
    position: int | None = None
    status: str = "draft"


class ParagraphUpdateRequest(BaseModel):
    kind: str | None = None
    position: int | None = None
    status: str | None = None


class ParagraphResponse(BaseModel):
    id: UUID
    project_id: UUID
    book_id: UUID
    chapter_id: UUID
    section_id: UUID
    kind: str
    position: int
    status: str
    word_count: int
    created_at: datetime
    updated_at: datetime
    sentence_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class SentenceCreateRequest(BaseModel):
    text: str = Field(min_length=1)
    kind: str = "body"
    position: int | None = None
    status: str = "draft"


class SentenceUpdateRequest(BaseModel):
    text: str | None = Field(default=None, min_length=1)
    kind: str | None = None
    position: int | None = None
    status: str | None = None


class SentenceResponse(BaseModel):
    id: UUID
    project_id: UUID
    book_id: UUID
    chapter_id: UUID
    section_id: UUID
    paragraph_id: UUID
    text: str
    kind: str
    position: int
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
